### PR TITLE
[PM-1880] Extract OrganizationService InviteUser(s)Async, SaveUserAsync and UpdateUserGroupsAsync to commands

### PR DESCRIPTION
**Pull request by r-tome**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Extract `OrganizationService` InviteUser(s)Async, SaveUserAsync and UpdateUserGroupsAsync to commands.

## Code changes

- **bitwarden_license/src/Commercial.Core/Services/ProviderService.cs:** Using `IInviteOrganizationUserCommand`
- **bitwarden_license/src/Scim/Users/PostUserCommand.cs:** Using `IInviteOrganizationUserCommand`
- **bitwarden_license/src/Scim/Utilities/ScimServiceCollectionExtensions.cs:** Adding `InviteOrganizationUserCommand` for dependency injection
- **bitwarden_license/test/Commercial.Core.Test/Services/ProviderServiceTests.cs:** Updated to use `IInviteOrganizationUserCommand`
- **bitwarden_license/test/Scim.Test/Users/PostUserCommandTests.cs:** Updated to use `IInviteOrganizationUserCommand`
- **src/Api/Controllers/OrganizationUsersController.cs:** Updated to use `ISaveOrganizationUserCommand`, `IUpdateOrganizationUserGroupsCommand` and `IInviteOrganizationUserCommand`
- **src/Api/Controllers/OrganizationsController.cs:** Using `IInviteOrganizationUserCommand`
- **src/Api/Public/Controllers/MembersController.cs:** Updated to use `ISaveOrganizationUserCommand`, `IUpdateOrganizationUserGroupsCommand` and `IInviteOrganizationUserCommand`
- **src/Api/Public/Controllers/OrganizationController.cs:** Using `IImportOrganizationCommand`
- **src/Core/OrganizationFeatures/Import/ImportOrganizationCommand.cs:** Extracted `ImportAsync` from `OrganizationService`
- **src/Core/OrganizationFeatures/Import/Interfaces/IImportOrganizationCommand.cs:** Command interface
- **src/Core/OrganizationFeatures/OrganizationServiceCollectionExtensions.cs:** Added new commands for dependency injection
- **src/Core/OrganizationFeatures/OrganizationUsers/Interfaces/IInviteOrganizationUserCommand.cs:** Command interface
- **src/Core/OrganizationFeatures/OrganizationUsers/Interfaces/ISaveOrganizationUserCommand.cs:** Command interface
- **src/Core/OrganizationFeatures/OrganizationUsers/Interfaces/IUpdateOrganizationUserGroupsCommand.cs:** Command interface
- **src/Core/OrganizationFeatures/OrganizationUsers/InviteOrganizationUserCommand.cs:** Extracted Invite methods from `OrganizationService`
- **src/Core/OrganizationFeatures/OrganizationUsers/OrganizationUserCommand.cs:** Base command containing reusable methods
- **src/Core/OrganizationFeatures/OrganizationUsers/SaveOrganizationUserCommand.cs:** Extracted `SaveUserAsync` from `OrganizationService`
- **src/Core/OrganizationFeatures/OrganizationUsers/UpdateOrganizationUserGroupsCommand.cs:** Extracted `UpdateUserGroupsAsync` from `OrganizationService`
- **src/Core/Services/IOrganizationService.cs:** Removed methods that were extracted into commands
- **src/Core/Services/Implementations/OrganizationService.cs:** Removed methods that were extracted into commands
- **test/Api.Test/Controllers/OrganizationsControllerTests.cs:** Using `IImportOrganizationCommand`
- **test/Core.Test/OrganizationFeatures/Import/ImportOrganizationCommandTests.cs:** Added unit tests for `ImportOrganizationCommand`
- **test/Core.Test/OrganizationFeatures/OrganizationUsers/InviteOrganizationUserCommandTests.cs:** Added unit tests for `InviteOrganizationUserCommand`
- **test/Core.Test/OrganizationFeatures/OrganizationUsers/SaveOrganizationUserCommandTests.cs:** Added unit tests for `SaveOrganizationUserCommand`
- **test/Core.Test/OrganizationFeatures/OrganizationUsers/UpdateOrganizationUserGroupsCommandTests.cs:** Added unit tests for `UpdateOrganizationUserGroupsCommand`
- **test/Core.Test/Services/OrganizationServiceTests.cs:** Removed tests for methods that were extracted

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by eliykat — CHANGES_REQUESTED** [review]
This looks great, thanks!

Is there any new or rewritten code in here, or is all just a lift & shift to the new commands, plus updating dependencies?

Please also resolve conflicts, thanks for your patience while I got to this review!

File: src/Core/OrganizationFeatures/OrganizationUsers/SaveOrganizationUserCommand.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        foreach (var kvp in permissionChecks)
+        {
+            var permissionName = kvp.Key;
+            var permissionCheckFunction = kvp.Value;
+
+            if (permissions.GetType().GetProperty(permissionName).GetValue(permissions) as bool? == true && !await permissionCheckFunction(user.OrganizationId))
+            {
+                throw new BadRequestException("Custom users can only grant the same custom permissions that they have.");
+            }
+        }
```
[the remark this answers is not in the document]
Comment (r-tome) [reply src/Core/OrganizationFeatures/OrganizationUsers/SaveOrganizationUserCommand.cs]: Yes, this was a merge issue. Its now fixed. Thanks for noticing

File: src/Core/OrganizationFeatures/OrganizationUsers/SaveOrganizationUserCommand.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+namespace Bit.Core.OrganizationFeatures.OrganizationUsers;
+
+public class SaveOrganizationUserCommand : OrganizationUserCommand, ISaveOrganizationUserCommand
+{
+    private readonly IEventService _eventService;
+    private readonly IOrganizationUserRepository _organizationUserRepository;
+    private readonly IOrganizationService _organizationService;
+
+    public SaveOrganizationUserCommand(
```
[the remark this answers is not in the document]
Comment (r-tome) [reply src/Core/OrganizationFeatures/OrganizationUsers/SaveOrganizationUserCommand.cs:18]: Makes sense!

File: src/Core/OrganizationFeatures/OrganizationUsers/InviteOrganizationUserCommand.cs [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    private async Task SendInvitesAsync(IEnumerable<OrganizationUser> orgUsers, Organization organization)
+    {
+        string MakeToken(OrganizationUser orgUser) =>
+            _dataProtector.Protect($"OrganizationUserInvite {orgUser.Id} {orgUser.Email} {CoreHelpers.ToEpocMilliseconds(DateTime.UtcNow)}");
+
+        await _mailService.BulkSendOrganizationInviteEmailAsync(organization.Name,
+            orgUsers.Select(o => (o, new ExpiringToken(MakeToken(o), DateTime.UtcNow.AddDays(5)))), organization.PlanType == PlanType.Free);
+    }
+
+    internal (bool canScale, string failureReason) CanScale(Organization organization,
```
[the remark this answers is not in the document]
Comment (r-tome) [reply src/Core/OrganizationFeatures/OrganizationUsers/InviteOrganizationUserCommand.cs]: I agree, `OrganizationService` has so much stuff in it that you can't see clearly what goes together.

File: src/Core/OrganizationFeatures/OrganizationUsers/OrganizationUserCommand.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
+﻿using Bit.Core.Context;
+using Bit.Core.Enums;
+using Bit.Core.Exceptions;
+using Bit.Core.Models.Data;
+using Bit.Core.Repositories;
+
+namespace Bit.Core.OrganizationFeatures.OrganizationUsers;
+
+public abstract class OrganizationUserCommand
```
[the remark this answers is not in the document]
Comment (r-tome) [reply src/Core/OrganizationFeatures/OrganizationUsers/OrganizationUserCommand.cs:9]: > This looks great, thanks!
> 
> Is there any new or rewritten code in here, or is all just a lift & shift to the new commands, plus updating dependencies?
> 
> Please also resolve conflicts, thanks for your patience while I got to this review!

Yes, its just lift and shift.
So many conflicts, it took me a while to fix all of it.

Thanks for reviewing it!

**Comment by eliykat** [comment]
@r-tome I think you've split this out into separate PRs now, should we close this one?

**Pull request closed without merging**